### PR TITLE
No longer compute checksum when installing images.

### LINKED ISSUE #10444: [1.5.0-rc] `docker push registry/app` insists "latest" tag must be present
Our builder uses source git SHAs as tags: we have no `latest` tag.

After upgrading one buildserver to `1.5.0`, I see this:

```
Jan 29 18:45:33.905:	time="2015-01-29T18:45:33Z" level="fatal" msg="Error response from daemon: Tag does not exist for localhost:5000/shopify: latest" 
```

For context, we are not doing, e.g. `docker push registry/app:1234deadbeefcafe` in this case, just `docker push registry/app`. We're just having it synchronize whatever it's got locally to the registry. 

#### Problem/Solution:

`docker push` with no tag specified should not insist on `latest` existing.

### PR DESCRIPTION
**Pull request by jlhawn**
While checksums are verified when a layer is pulled from v2 registries,
there are known issues where the checksum may change when the layer diff
is computed again. To avoid these issues, the checksum should no longer
be computed and stored until after it has been extracted to the docker
storage driver. The checksums are instead computed lazily before they
are pushed to a v2 registry.

Fixes #10444.

Docker-DCO-1.1-Signed-off-by: Josh Hawn josh.hawn@docker.com (github: jlhawn)

### CODE REVIEW

**Comment by jlhawn** [comment]
@tiborvass I've tried it out and this should fix the issue you reported here: https://github.com/docker/distribution/issues/112#issuecomment-71860918

**Comment by tiborvass** [comment]
Fixed the bug indeed.

LGTM

File: image/image.go [hunk 1]
```diff
@@ -80,48 +79,17 @@ func LoadImage(root string) (*Image, error) {
...
+func StoreImage(img *Image, layerData archive.ArchiveReader, root string) (err error) {
 	// Store the layer
-	var (
-		size        int64
-		err         error
-		driver      = img.graph.Driver()
-		layerTarSum tarsum.TarSum
-	)
-
+	var driver = img.graph.Driver()
```
Comment (crosbymichael) [inline image/image.go]: just a nit but `driver := img.graph.Driver()` never `var`

**Comment by jlhawn** [comment]
@tiborvass I just pushed again with the change suggested by @crosbymichael

**Comment by crosbymichael** [comment]
@jlhawn how does #10445 relate to this?  Is that other PR still needed?

**Comment by dmcgowan** [comment]
LGTM

**Comment by crosbymichael** [comment]
LGTM

**Comment by icecrime** [comment]
Closes #10444.

**Pull request merged**